File: Arrays/StockBuyAndSell.java
import java.util.*;
public class StockBuyAndSell {
    public static int brute(int prices[]){
        int profit=0;
        int n=prices.length;
        for(int i=0;i<n-1;i++){
            for(int j=i;j<n;j++){
                profit=Math.max(profit,prices[j]-prices[i]);
            }
        }
        return profit;
        //T.C=O(N^2)
    }
    public static int optimal(int prices[]){
        int n=prices.length;
        int min=prices[0];
        int profit=0;
        for(int i=1;i<n;i++){
            int cur=prices[i];
            if(min<cur){
                profit=Math.max(profit,cur-min);
            }
            else{
                min=cur;
            }
        }
        return profit;
        //T.C=O(N)
    }
    public static void main(String[] args) {
        int prices[]={17,20,11,9,12,6};
        System.out.println(brute(prices));
    } 
}

File: PYTHON/Arrays/StockBuyAndSell.py
def brute(prices):
    profit = 0
    n = len(prices)
    for i in range(n - 1):
        for j in range(i, n):
            profit = max(profit, prices[j] - prices[i])
    return profit  # T.C = O(N^2)
def optimal(prices):
    n = len(prices)
    min_price = prices[0]
    profit = 0
    for i in range(1, n):
        cur = prices[i]
        if min_price < cur:
            profit = max(profit, cur - min_price)
        else:
            min_price = cur
    return profit  # T.C = O(N)
if __name__ == "__main__":
    prices = [17, 20, 11, 9, 12, 6]
    print(brute(prices))  # Testing the brute force approach
    print(optimal(prices))  # Testing the optimal approach
